eliot/tests/test_action.py

```diff
@@ -1248,3 +1250,109 @@ def test_duplicate_task_level(self, start_message, child1, child2, index):
+            pass
+        extract_fields_for_failures(MyException,
+                                    lambda e: {"key": e.args[0]})
+        message = self.get_failed_action_message(MyException("a value"))
+        assertContainsFields(self, message, {"key": "a value"})
+
+    def test_subclass_falls_back_to_parent(self):
+        """
+        If an exception fails an action and the exact type has not been
+        registered but the error is a subclass of a registered class,
+        extract errors using it.
+        """
+        class MyException(Exception):
+            pass
+
+        class SubException(MyException):
+            pass
+
+        extract_fields_for_failures(MyException,
+                                    lambda e: {"key": e.args[0]})
+        message = self.get_failed_action_message(SubException("the value"))
+        assertContainsFields(self, message, {"key": "the value"})
+
+    def test_subclass_matches_first(self):
+        """
+        If both a superclass and base class have registered extractors, the
+        more specific one is used.
+        """
+        class MyException(Exception):
+            pass
+
+        class SubException(MyException):
+            pass
+
+        class SubSubException(SubException):
+            pass
+
+        extract_fields_for_failures(MyException,
+                                    lambda e: {"parent": e.args[0]})
+        extract_fields_for_failures(MyException,
+                                    lambda e: {"child": e.args[0]})
+        message = self.get_failed_action_message(SubSubException("the value"))
+        assertContainsFields(self, message, {"child": "the value"})
+
+    def test_error_in_extracter(self):
+        """
+        If an error occurs in extraction, log the message as usual just
+        without the extra fields, and an additional traceback.
+        """
+        class MyException(Exception):
+            pass
+
+        def extract(e):
+            return e.nosuchattribute
+        extract_fields_for_failures(MyException, extract)
+
+        logger = MemoryLogger()
+        action = Action(logger, "uuid", TaskLevel(level=[1]), "sys:me")
+        try:
+            with action:
+                raise MyException()
+        except MyException:
+            pass
+
+        assertContainsFields(self, logger.messages[1],
+                             {"action_type": "sys:me",
+                              "action_status": "failed"})
+        assertContainsFields(self, logger.messages[0],
+                             {"message_type": "eliot:traceback"})
+        self.assertIn("nosuchattribute", str(logger.messages[0]["reason"]))
+
+    def test_environmenterror(self):
+        """
+        ``EnvironmentError`` has a registered extractor that extracts the
+        errno.
+        """
+        message = self.get_failed_action_message(
+            EnvironmentError(12, "oh noes"))
+        assertContainsFields(self, message, {"errno": 12})
```

Comment:
Does this mean that we'll no longer see the message in the logs? Seems important to keep the message.